mne/time_frequency/psd.py

```diff
@@ -6,28 +6,22 @@
 
 from ..parallel import parallel_func
 from ..io.pick import _picks_to_idx
-from ..utils import logger, verbose, _time_mask
+from ..utils import logger, verbose, _time_mask, _check_option
 from .multitaper import psd_array_multitaper
 
 
-def _spect_func(epoch, n_overlap, n_per_seg, nfft, fs, freq_mask, func):
+def _spect_func(epoch, n_overlap, n_per_seg, nfft, fs, freq_mask, func,
+                average):
     """Aux function."""
     _, _, spect = func(epoch, fs=fs, nperseg=n_per_seg, noverlap=n_overlap,
                        nfft=nfft, window='hamming')
-    return spect[..., freq_mask, :]
-
-
-def _welch_func(epoch, n_overlap, n_per_seg, nfft, fs, freq_mask, average,
-                func):
-    """Aux function."""
-    kws = dict(fs=fs, nperseg=n_per_seg, noverlap=n_overlap, nfft=nfft,
-               window='hamming', average=average)
-
-    if average == 'mean':  # Compatibility with SciPy <1.2
-        del kws['average']
-
-    _, psd = func(epoch, **kws)
```

Comment:
... previously this func was `welch`, so within each `_welch_func` call the average across windows was taken. My commit preserves this behavior to save memory.

Reply:
But isn't the function parallelized over channels? And by default `n_jobs=1` so nothing runs in parallel.

Reply:
Whatever the parallelization is over, computing and averaging each within the loop then concatenating should consume less memory than computing each within a loop, concatenating, then averaging. Make sense?

Reply:
You mean potentially, right? Then yes, this makes sense. For `n_jobs=1` there shouldn't be a difference though.

Reply:
Even for one job the memory should be lower

Reply:
Before this PR (and after my commit in this PR), the operations (when `average is not None`) were:

1. For each channel/signal/whatever (parallelized or not):
    - Compute STFT to get something of shape `(n_freqs, n_segments)`
    - Average over windows/segments (these two steps were implicitly done by `welch` and in my commit are explicitly coded)
2. Concatenate results that are shape `(..., n_freqs)` to form the output

After your changes on top of `master` (before I pushed my commit) the operations were:

1. For each channel/signal/whatever (parallelized or not):
    - Compute STFT to get something of shape `(n_freqs, n_segments)`
2. Concatenate results that are shape `(..., n_freqs, n_segments`): requires memory of size `(..., n_freqs, n_segments)`, i.e. `n_segments` times more memory than step (2) above
3. Average over segments to obtain a result of shape `(..., n_freqs)` to form the output

Make sense?

Reply:
Maybe it's because it's already late, but no, I don't get it. Let's look at the default case of not parallelized, so `n_jobs=1` and `len(x_splits) = 1`. Then after your latest commit:
1. Compute spectrograms of shape `(n_channels, n_freqs, n_segments)`.
2. Average over segments to get shape `(n_channels, n_freqs)`.
3. Concatenate results, which is a no-op because `len(f_spect) = 1`.

My version:
1. Compute spectrograms of shape `(n_channels, n_freqs, n_segments)`.
2. Concatenate results, which is a no-op because `len(f_spect) = 1`.
3. Average over segments to get shape `(n_channels, n_freqs)`.

So in both versions we go from `(n_channels, n_freqs, n_segments)` to `(n_channels, n_freqs)` with basically the same steps because we're not concatenating anything.

Reply:
Ahh you're right, it will only matter in the parallelized case currently. I hadn't noticed that the parallelization was over an `array_split` of X and not just X itself. So I was wrong about the dimensions here -- good catch!

However, `np.concatenate(...)` always makes a copy, so even if it's just a list with a single element it's not a no-op, it copies. So there still will be a factor of 2x memory pain from not doing it inside the loop -- but at least it's much less than I thought.

FWIW we should actually probably change the parallelization behavior since, unlike in some of our functions, there won't be any benefit here of iterating over the `array_split`, and there is a (potentially quite large) memory cost to doing it this way.

Feel free to rebase and then this should be good to go. I say we keep my version so that we at least get the 2x (plus more in parallelization) benefit in the current case.

Reply:
> Ahh you're right, it will only matter in the parallelized case currently. I hadn't noticed that the parallelization was over an `array_split` of X and not just X itself. So I was wrong about the dimensions here -- good catch!

😌 - took me some time to figure it out and I'm still not sure I got it all sorted out (see below).

> However, `np.concatenate(...)` always makes a copy, so even if it's just a list with a single element it's not a no-op, it copies. So there still will be a factor of 2x memory pain from not doing it inside the loop -- but at least it's much less than I thought.

I think it's actually not just 2x, but `n_segments`x. In your version, we concatenate (make a copy of) `[(n_channels, n_freqs)]` and in my version we concatenate (make a copy of) `[(n_channels, n_freqs, n_segments)]`. So we should definitely keep your version.

> FWIW we should actually probably change the parallelization behavior since, unlike in some of our functions, there won't be any benefit here of iterating over the `array_split`, and there is a (potentially quite large) memory cost to doing it this way.

I was wondering why we need the parallelization at all here. `scipy.signal.spectrogram` is already vectorized over channels/all dimensions except the one we compute over (I assume), so what is the use case for splitting up? And what is split up?